compiler/rustc_parse/src/lexer/mod.rs

```diff
@@ -137,12 +137,45 @@ impl<'a> StringReader<'a> {
         // Opening delimiter of the length 2 is not included into the comment text.
         let content_start = start + BytePos(2);
         let content = self.str_from(content_start);
-        let span = self.mk_sp(start, self.pos);
-        const UNICODE_TEXT_FLOW_CHARS: &[char] = &[
-            '\u{202A}', '\u{202B}', '\u{202D}', '\u{202E}', '\u{2066}', '\u{2067}', '\u{2068}',
-            '\u{202C}', '\u{2069}',
-        ];
-        if content.contains(UNICODE_TEXT_FLOW_CHARS) {
+
+        // Char   - UTF-8
+        // U+202A - E2 80 AA
+        // U+202B - E2 80 AB
+        // U+202C - E2 80 AC
+        // U+202D - E2 80 AD
+        // U+202E - E2 80 AE
+        // U+2066 - E2 81 A6
+        // U+2067 - E2 81 A7
+        // U+2068 - E2 81 A8
+        // U+2069 - E2 81 A9
+        let mut bytes = content.as_bytes();
+        let contains_flow_control_chars = loop {
+            match core::slice::memchr::memchr(0xE2, &bytes) {
+                Some(idx) => {
+                    // bytes are valid UTF-8 -> E2 must be followed by two bytes
+                    match bytes[idx + 1] {
+                        0x80 => {
+                            if (0xAA..=0xAE).contains(&bytes[idx + 2]) {
```

Comment:
Given that indexing includes a bounds check, I wonder if the compiler will be able to optimize these into a single test, rather than two separate tests. It doesn't matter much, since the appearance of 0xE2 will already make this the uncommon case, but it may become an issue if people start writing other Unicode that includes 0xE2.

To give the compiler the best chance of optimizing this, how about starting the `Some` case above with `let ch = bytes[idx..=idx+2];` (right after the comment explaining why that's correct), and then indexing into `ch`? I think that'll make the code a little clearer regardless.

Reply:
`let ch = &bytes[idx..=idx+2];` compiles with optimization to _three_ comparisons to differentiate between those conditions: `idx+2 != usize::max` (for panicking with "attempted to index slice up to maximum usize"), `idx<=idx+2` (not optimized away because the addtion could overflow) and `idx + 2 < len`. We could get rid of the first one using `let ch = &bytes[idx..idx+3];` but that would still mean one more comparison for the common case that second byte is neither 0x80 nor 0x81. ([Godbolt](https://godbolt.org/z/cdK4ejjq7))

Reply:
The code size is quite a bit smaller though, so it might be worth it anyway.

Reply:
Does evaluating the bytes separate improve performance compared to

```rust
match ch {
   [_, 0x80, 0xAA..=0xAE] | [_, 0x81, 0xA6..=0xA9] => break true,
   _ => {}
}
```

Reply:
Evaluating the bytes separately made a difference before without the subslice because one overflow check could be delayed. Now all checks are done at subslice creation so the assembly is the same (modulo reordering). I will change it.